nameresolution/consul/watcher.go

```diff
@@ -0,0 +1,342 @@
+		return waitIndexVal(0)
+	}
+
+	return idx
+}
+
+type serviceIdentifier struct {
+	serviceName string
+	serviceID   string
+	node        string
+}
+
+func getServiceIdentifier(c *consul.HealthCheck) serviceIdentifier {
+	return serviceIdentifier{
+		serviceID:   c.ServiceID,
+		serviceName: c.ServiceName,
+		node:        c.Node}
+}
+
+func getHealthByService(checks consul.HealthChecks) map[serviceIdentifier]bool {
+	healthByService := make(map[serviceIdentifier]bool)
+	for _, check := range checks {
+		id := getServiceIdentifier(check)
+
+		if state, ok := healthByService[id]; !ok {
+			// Init to healthy
+			healthByService[id] = true
+		} else if !state {
+			continue
+		}
+
+		if check.Status != consul.HealthPassing {
+			healthByService[id] = false
+		}
+	}
+
+	return healthByService
+}
+
+func (p *watchPlan) getChangedServices(newResult map[serviceIdentifier]bool) map[string]struct{} {
+	changedServices := make(map[string]struct{})
+
+	// get changed services
+	for newKey, newValue := range newResult {
+		if oldValue, ok := p.lastResult[newKey]; ok && newValue == oldValue {
+			continue
+		}
+
+		changedServices[newKey.serviceName] = struct{}{}
+	}
+
+	for oldKey := range p.lastResult {
+		if _, ok := newResult[oldKey]; !ok {
+			changedServices[oldKey.serviceName] = struct{}{}
+		}
+	}
+
+	return changedServices
+}
+
+func (p *watchPlan) buildServiceNameFilter(services []string) {
+	var nameFilters = make([]string, len(services))
+
+	for i, v := range services {
+		nameFilters[i] = fmt.Sprintf("ServiceName==\"%s\"", v)
+	}
+
+	filter := fmt.Sprintf("(%s)", strings.Join(nameFilters, " or "))
+
+	if len(p.configuredQueryFilter) < 1 {
+		p.options.Filter = filter
+	} else {
+		p.options.Filter = fmt.Sprintf("%s and %s", p.configuredQueryFilter, filter)
+	}
+}
+
+func (r *resolver) watch(p *watchPlan, services []string, ctx context.Context) (blockingParamVal, consul.HealthChecks, error, bool) {
+	p.options = p.options.WithContext(ctx)
+
+	if p.lastParamVal != nil {
+		p.options.WaitIndex = uint64(p.lastParamVal.(waitIndexVal))
+	}
+
+	p.buildServiceNameFilter(services)
+
+	checks, meta, err := r.client.Health().State(consul.HealthAny, p.options)
+
+	if err != nil {
+		// if it failed during long poll try again with no wait
+		if p.options.WaitIndex != uint64(0) {
+			p.options.WaitIndex = 0
+			checks, meta, err = r.client.Health().State(consul.HealthAny, p.options)
+		}
+
+		if err != nil {
+			// if the context was canceled
+			if errors.Is(err, context.Canceled) {
+				return nil, nil, nil, true
+			}
+
+			// if it failed with no wait and plan is not expired
+			if p.options.WaitIndex == uint64(0) && !p.expired {
+				p.lastResult = nil
+				p.expired = true
+				r.registry.expireAll()
+			}
+
+			return nil, nil, err, false
+		}
+	}
+
+	p.expired = false
+
+	return waitIndexVal(meta.LastIndex), checks, err, false
+}
+
+func (r *resolver) runWatchPlan(p *watchPlan, services []string, ctx context.Context, watchTask chan bool) {
```

Comment:
`ctx` must always be the first argument in a function signature: https://pkg.go.dev/context

> The Context should be the first parameter, typically named ctx